src/util/elf_parser.cpp

```diff
@@ -0,0 +1,300 @@
+#include <algorithm>
+#include <cerrno>
+#include <cstring>
+#include <elf.h>
+#include <fcntl.h>
+#include <filesystem>
+#include <gelf.h>
+#include <iostream>
+#include <libelf.h>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <unistd.h>
+#include <unordered_map>
+#include <vector>
+
+#include "log.h"
+#include "util/elf_parser.h"
+#include "util/hash.h"
+
+namespace bpftrace::util {
+
+using namespace std::string_literals;
+
+char ELFParseError::ID;
+
+void ELFParseError::log(llvm::raw_ostream& OS) const
+{
+  OS << msg_;
+}
+
+struct PairHash {
+  std::size_t operator()(const std::pair<std::string, std::string>& p) const
+  {
+    std::size_t seed = 0;
+    hash_combine(seed, p.first);
+    hash_combine(seed, p.second);
+    return seed;
+  }
+};
+
+// ELFParser implementation
+
+ELFParser::ELFParser(std::string path, Elf* elf)
+    : path_(std::move(path)), elf_(elf)
+{
+}
+
+Result<std::vector<struct elf_segment>> ELFParser::parse_segments()
+{
+  std::vector<struct elf_segment> segs;
+  GElf_Phdr phdr;
+  size_t n;
+
+  if (elf_getphdrnum(elf_, &n)) {
+    return make_error<ELFParseError>(
+        "usdt: failed to process ELF program segments for '"s + path_ +
+        "': " + strerror(-errno));
+  }
+
+  for (size_t i = 0; i < n; i++) {
+    if (!gelf_getphdr(elf_, i, &phdr)) {
+      using namespace std::string_literals;
+      return make_error<ELFParseError>(
+          "usdt: failed to process ELF program segments for '"s + path_ +
+          "': " + strerror(-errno));
+    }
+
+    if (phdr.p_type != PT_LOAD)
+      continue;
+    segs.emplace_back(static_cast<long>(phdr.p_vaddr),
+                      static_cast<long>(phdr.p_vaddr + phdr.p_memsz),
+                      static_cast<long>(phdr.p_offset),
+                      static_cast<bool>(phdr.p_flags & PF_X));
+  }
+
+  if (segs.empty()) {
+    LOG(WARNING) << "failed to find PT_LOAD program headers in '" << path_;
+  }
+
+  std::ranges::sort(segs, [](const elf_segment& a, const elf_segment& b) {
+    return a.start < b.start;
+  });
+  return segs;
+}
+
+Result<std::pair<Elf_Scn*, GElf_Shdr>> ELFParser::find_section_by_name(
+    std::string_view sec_name)
+{
+  Elf_Scn* sec = nullptr;
+  size_t shstrndx;
+  GElf_Shdr shdr;
+
+  // get elf string table index
+  if (elf_getshdrstrndx(elf_, &shstrndx)) {
+    return make_error<ELFParseError>("failed to get string table index from " +
+                                     path_ + ": " + elf_errmsg(-1));
+  }
+
+  /* check if ELF is corrupted and avoid calling elf_strptr if yes */
```

Comment:
Nit: we tend to use C++ style comments everywhere

```suggestion
  // check if ELF is corrupted and avoid calling elf_strptr if yes
```